File: sdk/go/mockagents/mcp_test.go
package mockagents

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"
	"time"
)

// newMcpTestServer spins up a canned httptest.Server that serves
// GET /mcp/events with the supplied SSE frames and collects every
// POST /mcp/response body for later inspection.
func newMcpTestServer(t *testing.T, frames []string) (*httptest.Server, *[]string) {
	t.Helper()
	var (
		mu        sync.Mutex
		responses []string
	)
	mux := http.NewServeMux()
	mux.HandleFunc("GET /mcp/events", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		w.WriteHeader(200)
		flusher, _ := w.(http.Flusher)
		for _, frame := range frames {
			_, _ = fmt.Fprint(w, frame+"\n\n")
			if flusher != nil {
				flusher.Flush()
			}
		}
	})
	mux.HandleFunc("POST /mcp/response", func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		mu.Lock()
		responses = append(responses, string(body))
		mu.Unlock()
		w.WriteHeader(http.StatusAccepted)
	})
	srv := httptest.NewServer(mux)
	t.Cleanup(srv.Close)
	return srv, &responses
}

func TestMcpClient_ParsesRequestFrame(t *testing.T) {
	frames := []string{
		`event: request
data: {"jsonrpc":"2.0","id":1,"method":"sampling/createMessage","params":{"prompt":"hi"}}`,
	}
	srv, _ := newMcpTestServer(t, frames)

	client := NewMcpClient(McpClientOptions{BaseURL: srv.URL})
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	stream, err := client.Connect(ctx)
	if err != nil {
		t.Fatalf("Connect: %v", err)
	}
	defer stream.Close()

	if !stream.Next() {
		t.Fatalf("Next returned false: %v", stream.Err())
	}
	ev := stream.Value()
	if !ev.IsRequest() {
		t.Errorf("expected request, kind = %q", ev.Kind)
	}
	if ev.Payload.Method != "sampling/createMessage" {
		t.Errorf("method = %q", ev.Payload.Method)
	}
	if ev.Params()["prompt"] != "hi" {
		t.Errorf("params = %+v", ev.Params())
	}
}

func TestMcpClient_SkipsHeartbeatsAndMalformed(t *testing.T) {
	frames := []string{
		`:heartbeat`,
		`event: request
data: not-json`,
		`event: request
data: {"jsonrpc":"2.0","id":2,"method":"roots/list"}`,
	}
	srv, _ := newMcpTestServer(t, frames)

	client := NewMcpClient(McpClientOptions{BaseURL: srv.URL})
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	stream, err := client.Connect(ctx)
	if err != nil {
		t.Fatal(err)
	}
	defer stream.Close()

	var methods []string
	for stream.Next() {
		methods = append(methods, stream.Value().Payload.Method)
	}
	if len(methods) != 1 || methods[0] != "roots/list" {
		t.Errorf("methods = %v", methods)
	}
}

func TestMcpClient_SendResponseResult(t *testing.T) {
	srv, recorded := newMcpTestServer(t, nil)
	client := NewMcpClient(McpClientOptions{BaseURL: srv.URL})

	err := client.SendResponse(context.Background(),
		json.RawMessage("7"),
		map[string]any{"text": "pong"},
		nil,
	)
	if err != nil {
		t.Fatalf("SendResponse: %v", err)
	}
	if len(*recorded) != 1 {
		t.Fatalf("recorded = %d", len(*recorded))
	}
	var body map[string]any
	if err := json.Unmarshal([]byte((*recorded)[0]), &body); err != nil {
		t.Fatal(err)
	}
	if body["result"].(map[string]any)["text"] != "pong" {
		t.Errorf("body = %+v", body)
	}
}

func TestMcpClient_SendResponseError(t *testing.T) {
	srv, recorded := newMcpTestServer(t, nil)
	client := NewMcpClient(McpClientOptions{BaseURL: srv.URL})

	err := client.SendResponse(context.Background(),
		json.RawMessage(`"abc"`),
		nil,
		&JSONRPCError{Code: -32603, Message: "boom"},
	)
	if err != nil {
		t.Fatal(err)
	}
	var body map[string]any
	_ = json.Unmarshal([]byte((*recorded)[0]), &body)
	errObj := body["error"].(map[string]any)
	if errObj["code"].(float64) != -32603 {
		t.Errorf("code = %v", errObj["code"])
	}
}

func TestMcpClient_SendResponseRequiresExactlyOneOf(t *testing.T) {
	client := NewMcpClient(McpClientOptions{BaseURL: "http://nowhere"})
	err := client.SendResponse(context.Background(), json.RawMessage("1"), nil, nil)
	if err == nil {
		t.Fatal("expected error")
	}
	err = client.SendResponse(context.Background(), json.RawMessage("1"),
		map[string]any{}, &JSONRPCError{Code: 1, Message: "x"})
	if err == nil {
		t.Fatal("expected error for both supplied")
	}
}

func TestMcpClient_DispatchRequestRoutesToHandler(t *testing.T) {
	srv, recorded := newMcpTestServer(t, nil)
	client := NewMcpClient(McpClientOptions{BaseURL: srv.URL})

	event := &McpEvent{
		Kind: "request",
		Payload: JSONRPCEnvelope{
			JSONRPC: "2.0",
			ID:      json.RawMessage("42"),
			Method:  "sampling/createMessage",
			Params:  map[string]any{"prompt": "hi"},
		},
	}

	var seen map[string]any
	handlers := map[string]McpRequestHandler{
		"sampling/createMessage": func(_ context.Context, params map[string]any) (map[string]any, error) {
			seen = params
			return map[string]any{"text": "ok"}, nil
		},
	}
	result, err := client.DispatchRequest(context.Background(), event, handlers)
	if err != nil {
		t.Fatal(err)
	}
	if result["text"] != "ok" || seen["prompt"] != "hi" {
		t.Errorf("result=%+v seen=%+v", result, seen)
	}
	var body map[string]any
	_ = json.Unmarshal([]byte((*recorded)[0]), &body)
	if body["result"].(map[string]any)["text"] != "ok" {
		t.Errorf("posted body = %+v", body)
	}
}

func TestMcpClient_DispatchRequestUnknownMethod(t *testing.T) {
	srv, recorded := newMcpTestServer(t, nil)
	client := NewMcpClient(McpClientOptions{BaseURL: srv.URL})
	event := &McpEvent{
		Kind: "request",
		Payload: JSONRPCEnvelope{
			ID:     json.RawMessage("1"),
			Method: "unknown/method",
		},
	}
	_, err := client.DispatchRequest(context.Background(), event, map[string]McpRequestHandler{})
	if err == nil {
		t.Fatal("expected error")
	}
	// The client should have posted a -32601 error.
	var body map[string]any
	_ = json.Unmarshal([]byte((*recorded)[0]), &body)
	errObj := body["error"].(map[string]any)
	if errObj["code"].(float64) != -32601 {
		t.Errorf("code = %v", errObj["code"])
	}
}

func TestMcpClient_DispatchRequestHandlerErrorPostsInternal(t *testing.T) {
	srv, recorded := newMcpTestServer(t, nil)
	client := NewMcpClient(McpClientOptions{BaseURL: srv.URL})
	event := &McpEvent{
		Kind: "request",
		Payload: JSONRPCEnvelope{
			ID:     json.RawMessage("9"),
			Method: "sampling/createMessage",
		},
	}
	handlers := map[string]McpRequestHandler{
		"sampling/createMessage": func(_ context.Context, _ map[string]any) (map[string]any, error) {
			return nil, errors.New("kaboom")
		},
	}
	_, err := client.DispatchRequest(context.Background(), event, handlers)
	if err == nil || err.Error() != "kaboom" {
		t.Errorf("err = %v", err)
	}
	var body map[string]any
	_ = json.Unmarshal([]byte((*recorded)[0]), &body)
	errObj := body["error"].(map[string]any)
	if errObj["code"].(float64) != -32603 {
		t.Errorf("code = %v", errObj["code"])
	}
	if errObj["message"] != "kaboom" {
		t.Errorf("message = %v", errObj["message"])
	}
}

func TestMcpClient_DispatchRequestRejectsNotification(t *testing.T) {
	client := NewMcpClient(McpClientOptions{BaseURL: "http://nowhere"})
	event := &McpEvent{
		Kind: "notification",
		Payload: JSONRPCEnvelope{
			Method: "notifications/tools/list_changed",
		},
	}
	_, err := client.DispatchRequest(context.Background(), event, nil)
	if err == nil {
		t.Fatal("expected error")
	}
}

func TestMcpEvent_IsRequestRequiresID(t *testing.T) {
	withID := &McpEvent{Kind: "request", Payload: JSONRPCEnvelope{ID: json.RawMessage("1")}}
	if !withID.IsRequest() {
		t.Error("event with id should be a request")
	}
	withoutID := &McpEvent{Kind: "request", Payload: JSONRPCEnvelope{Method: "m"}}
	if withoutID.IsRequest() {
		t.Error("event without id should not be a request")
	}
	notif := &McpEvent{Kind: "notification", Payload: JSONRPCEnvelope{Method: "m"}}
	if notif.IsRequest() {
		t.Error("notification should not be a request")
	}
}

func TestMcpEvent_ParamsAlwaysReturnsMap(t *testing.T) {
	ev := &McpEvent{Payload: JSONRPCEnvelope{Method: "m"}}
	if ev.Params() == nil {
		t.Error("Params() must be non-nil")
	}
	if len(ev.Params()) != 0 {
		t.Errorf("Params() = %+v", ev.Params())
	}
}

File: sdk/typescript/tests/mcp.test.ts
// MCP bidirectional helper tests. Like streaming.test.ts these run
// against an in-process node:http server that serves canned SSE
// frames on GET /mcp/events and collects POST /mcp/response bodies
// for later inspection.

import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  expect as vexpect,
  it,
} from "vitest";
import { createServer, Server, IncomingMessage, ServerResponse } from "node:http";
import { AddressInfo } from "node:net";

import {
  McpClient,
  McpEvent,
  isRequest,
  paramsOf,
  parseMcpFrame,
} from "../src/mcp.js";

interface Recorded {
  method: string;
  url: string;
  body: string;
}

let server: Server;
let port: number;

// Canned /mcp/events frames. Each test can overwrite this before
// calling connect() to drive a different scenario.
let eventsFrames: string[] = [];

// Every POST /mcp/response body we saw during the test.
let recorded: Recorded[] = [];

// When set, the /mcp/response endpoint returns this status instead
// of 202 — used to exercise the error path.
let responseStatus = 202;

function writeSSE(res: ServerResponse, frames: string[]) {
  res.writeHead(200, {
    "Content-Type": "text/event-stream",
    "Cache-Control": "no-cache",
    Connection: "keep-alive",
  });
  for (const frame of frames) {
    res.write(frame + "\n\n");
  }
  res.end();
}

beforeAll(async () => {
  server = createServer((req: IncomingMessage, res: ServerResponse) => {
    if (req.method === "GET" && req.url === "/mcp/events") {
      writeSSE(res, eventsFrames);
      return;
    }
    if (req.method === "POST" && req.url === "/mcp/response") {
      let body = "";
      req.on("data", (chunk) => (body += chunk.toString()));
      req.on("end", () => {
        recorded.push({ method: req.method!, url: req.url!, body });
        res.writeHead(responseStatus);
        res.end();
      });
      return;
    }
    res.writeHead(404);
    res.end("not found");
  });
  await new Promise<void>((resolve) => server.listen(0, resolve));
  port = (server.address() as AddressInfo).port;
});

afterAll(async () => {
  await new Promise<void>((resolve) => server.close(() => resolve()));
});

beforeEach(() => {
  eventsFrames = [];
  recorded = [];
  responseStatus = 202;
});

function baseUrl(): string {
  return `http://127.0.0.1:${port}`;
}

async function collectEvents(client: McpClient): Promise<McpEvent[]> {
  const stream = client.connect();
  const out: McpEvent[] = [];
  for await (const event of stream) {
    out.push(event);
  }
  await stream.close();
  return out;
}

// --- parseMcpFrame unit tests (no HTTP) ---

describe("parseMcpFrame", () => {
  it("decodes a request frame", () => {
    const ev = parseMcpFrame(
      'event: request\ndata: {"jsonrpc":"2.0","id":1,"method":"sampling/createMessage","params":{"prompt":"hi"}}',
    );
    vexpect(ev).not.toBeNull();
    vexpect(ev!.kind).toBe("request");
    vexpect(ev!.payload.id).toBe(1);
    vexpect(ev!.payload.method).toBe("sampling/createMessage");
  });

  it("decodes a notification frame", () => {
    const ev = parseMcpFrame(
      'event: notification\ndata: {"jsonrpc":"2.0","method":"notifications/tools/list_changed"}',
    );
    vexpect(ev).not.toBeNull();
    vexpect(ev!.kind).toBe("notification");
    vexpect(ev!.payload.id).toBeUndefined();
  });

  it("drops frames with malformed data", () => {
    const ev = parseMcpFrame("event: request\ndata: not-json");
    vexpect(ev).toBeNull();
  });

  it("drops frames with no data line", () => {
    const ev = parseMcpFrame("event: request\n");
    vexpect(ev).toBeNull();
  });
});

// --- isRequest / paramsOf accessors ---

describe("McpEvent accessors", () => {
  it("isRequest true only with an id", () => {
    vexpect(
      isRequest({
        kind: "request",
        payload: { jsonrpc: "2.0", id: 1, method: "m" },
      }),
    ).toBe(true);
    vexpect(
      isRequest({ kind: "request", payload: { jsonrpc: "2.0", method: "m" } }),
    ).toBe(false);
    vexpect(
      isRequest({
        kind: "notification",
        payload: { jsonrpc: "2.0", method: "m" },
      }),
    ).toBe(false);
  });

  it("paramsOf always returns a dict", () => {
    vexpect(
      paramsOf({ kind: "request", payload: { jsonrpc: "2.0", method: "m" } }),
    ).toEqual({});
    vexpect(
      paramsOf({
        kind: "request",
        payload: { jsonrpc: "2.0", method: "m", params: ["oops"] },
      }),
    ).toEqual({});
    vexpect(
      paramsOf({
        kind: "request",
        payload: { jsonrpc: "2.0", method: "m", params: { a: 1 } },
      }),
    ).toEqual({ a: 1 });
  });
});

// --- end-to-end against the canned server ---

describe("McpClient against node:http server", () => {
  it("parses a request frame from /mcp/events", async () => {
    eventsFrames = [
      'event: request\ndata: {"jsonrpc":"2.0","id":1,"method":"sampling/createMessage","params":{"prompt":"hi"}}',
    ];
    const client = new McpClient({ baseUrl: baseUrl() });
    const events = await collectEvents(client);
    vexpect(events).toHaveLength(1);
    vexpect(events[0].kind).toBe("request");
    vexpect(events[0].payload.method).toBe("sampling/createMessage");
  });

  it("skips heartbeats and malformed frames", async () => {
    eventsFrames = [
      ":heartbeat",
      "event: request\ndata: not-json",
      'event: request\ndata: {"jsonrpc":"2.0","id":2,"method":"roots/list"}',
    ];
    const client = new McpClient({ baseUrl: baseUrl() });
    const events = await collectEvents(client);
    vexpect(events).toHaveLength(1);
    vexpect(events[0].payload.method).toBe("roots/list");
  });

  it("sendResponse posts the expected JSON body", async () => {
    const client = new McpClient({ baseUrl: baseUrl() });
    await client.sendResponse(7, { result: { text: "pong" } });
    vexpect(recorded).toHaveLength(1);
    const body = JSON.parse(recorded[0].body);
    vexpect(body).toEqual({
      jsonrpc: "2.0",
      id: 7,
      result: { text: "pong" },
    });
  });

  it("sendResponse rejects when neither result nor error is supplied", async () => {
    const client = new McpClient({ baseUrl: baseUrl() });
    await vexpect(client.sendResponse(1, {})).rejects.toThrow(/exactly one/);
  });

  it("sendResponse throws HTTPError on non-2xx", async () => {
    responseStatus = 404;
    const client = new McpClient({ baseUrl: baseUrl() });
    await vexpect(
      client.sendResponse(999, { result: {} }),
    ).rejects.toThrow(/404/);
  });

  it("dispatchRequest routes to handler and posts the result", async () => {
    const client = new McpClient({ baseUrl: baseUrl() });
    const event: McpEvent = {
      kind: "request",
      payload: {
        jsonrpc: "2.0",
        id: 42,
        method: "sampling/createMessage",
        params: { prompt: "hi" },
      },
    };
    let seen: Record<string, unknown> | null = null;
    const result = await client.dispatchRequest(event, {
      "sampling/createMessage": (params) => {
        seen = params;
        return { text: "ok" };
      },
    });
    vexpect(result).toEqual({ text: "ok" });
    vexpect(seen).toEqual({ prompt: "hi" });
    vexpect(recorded).toHaveLength(1);
    const body = JSON.parse(recorded[0].body);
    vexpect(body.id).toBe(42);
    vexpect(body.result).toEqual({ text: "ok" });
  });

  it("dispatchRequest posts -32601 for unknown method and throws", async () => {
    const client = new McpClient({ baseUrl: baseUrl() });
    const event: McpEvent = {
      kind: "request",
      payload: { jsonrpc: "2.0", id: 1, method: "unknown/method", params: {} },
    };
    await vexpect(client.dispatchRequest(event, {})).rejects.toThrow(
      "unknown/method",
    );
    const body = JSON.parse(recorded[0].body);
    vexpect(body.error.code).toBe(-32601);
  });

  it("dispatchRequest posts -32603 and rethrows when the handler throws", async () => {
    const client = new McpClient({ baseUrl: baseUrl() });
    const event: McpEvent = {
      kind: "request",
      payload: { jsonrpc: "2.0", id: 9, method: "sampling/createMessage", params: {} },
    };
    await vexpect(
      client.dispatchRequest(event, {
        "sampling/createMessage": () => {
          throw new Error("kaboom");
        },
      }),
    ).rejects.toThrow("kaboom");
    const body = JSON.parse(recorded[0].body);
    vexpect(body.error.code).toBe(-32603);
    vexpect(body.error.message).toContain("kaboom");
  });

  it("dispatchRequest rejects non-request events", async () => {
    const client = new McpClient({ baseUrl: baseUrl() });
    const event: McpEvent = {
      kind: "notification",
      payload: { jsonrpc: "2.0", method: "notifications/tools/list_changed" },
    };
    await vexpect(client.dispatchRequest(event, {})).rejects.toThrow(/non-request/);
  });
});
